File: php/src/HelloWorld.php
<?php

namespace App;

class HelloWorld
{
    public function sayHello(): string
    {
        $message = "Hello, SonarQube PHP!";
        
        // Unused variable
        $greeting = $this->getGreeting();
        
        return $message;
    }

    public function greet(string $name): string
    {
        // Unused variable
        $formattedName = trim($name);
        
        return "Hello, $name!";
    }

    private function getGreeting(): string
    {
        // Dead code - unused private method
        return "Greetings!";
    }

    private function formatOutput(string $text): string
    {
        // Unused method
        return strtoupper($text);
    }
}

echo (new HelloWorld())->sayHello() . "\n";

File: ruby/src/HelloWorld.rb
class HelloWorld
  # Code smell: class variable without @@
  count = 0

  def initialize
    @greeting = 'Hello, World!'
  end

  def say
    puts @greeting
  end

  def self.greeting
    # Code smell: instance variable without @
    message = 'Hello from class method!'
    puts message
    message
  end

  def format_output(text)
    # Duplicate code: similar formatting
    "[OUTPUT] #{text}"
  end

  def format_result(result)
    # Duplicate code: similar formatting
    "[RESULT] #{result}"
  end
end

if __FILE__ == $0
  hello = HelloWorld.new
  hello.say
  HelloWorld.greeting
end